### ISSUE TITLE
Expose service of native cache library

### ISSUE THREAD

**Issue by artembilan** [spring-projects/spring-boot#2848]
The new `HazelcastCacheConfiguration` doesn't utilize `HazelcastInstance` as a `@Bean`, however it won't help, because this `HazelcastCacheConfiguration` is tied with a Cache abstraction anyway.

Some other projects, like Spring Session, Spring Integration, Spring Cloud really need `HazelcastInstance` for their purposes, even if that `auto-config` will just utilize the picking up for the `Config`

**Comment by jvalkeal** [comment]
I think we need to harmonize this with SI and Spring Cloud. Although WIP at this moment in [HazelcastLeaderAutoConfiguration](https://github.com/spring-cloud/spring-cloud-cluster/blob/master/spring-cloud-cluster-autoconfigure/src/main/java/org/springframework/cloud/cluster/autoconfigure/leader/HazelcastLeaderAutoConfiguration.java) we do

```java
@Bean
public HazelcastInstance hazelcastInstance() {
	return Hazelcast.newHazelcastInstance(hazelcastConfig());
}
```
So it'd be a great if we can do this in a central point or use a same config tricks.

**Comment by artembilan** [comment]
@snicoll , thank you for the quick acceptance. If you want to do this yourself, pay attention, please, to the `shutdown()` method. The `HazelcastInstance` bean must be like this:
````java
@Bean(destroyMethod = "shutdown")
public HazelcastInstance hazelcastInstance() {
	return Hazelcast.newHazelcastInstance(hazelcastConfig());
}
````
Because our Spring App can be as a _deployable_ app and our JVM might not be terminated, like the default Hazelcast behaviour is done. See `com.hazelcast.instance.Node#start()`.

**Comment by snicoll** [comment]
I know that and that is also something that bothers me actually.

You may also configure a `HazelcastInstance` that is shared with another application and stopping your app should not terminate the cache manager. I am not sure how bad it is.

Currently, if you expose a bean using Java config and it has either a `close` or `shutdown` method, it is automatically registered (and there is no way to unregister it). I'd argue it would be the user's responsibility to figure out what he wants.

As a final note, if you need the native `CacheManager`, you should be able to access it via the cache abstraction. Unfortunately, the hazelcast implementation does not expose it, see https://github.com/hazelcast/hazelcast/issues/4978

**Comment by artembilan** [comment]
OK. I see `DisposableBeanAdapter#inferDestroyMethodIfNecessary`. Thank you! Didn't know that and saw Hazelcast members overhead with XML configs in our test suite until I haven't provided `destroy-method="shutdown"`. That's why that "infer" feature for Java config has bypassed me :smile: : http://docs.spring.io/spring/docs/current/spring-framework-reference/htmlsingle/#beans-factory-lifecycle-disposablebean.

I'd say that it would be really bad to share `instance` (Hazelcast member) between apps. Although that can be possible only if we use the `instanceName` explicitly. But, frankly, it's up to end-user finally:

````java
@Bean
public HazelcastInstance hazelcastInstance() {
	return Hazelcast.getOrCreateHazelcastInstance(new Config().setInstanceName("foo"));
}

@Bean
public HazelcastInstance hazelcastInstance2() {
	return Hazelcast.getOrCreateHazelcastInstance(new Config().setInstanceName("foo"));
}
````
As you see the `instanceName` is a part of Hazelcast config, so to be flexible we should use `getOrCreateHazelcastInstance` factory method in our auto-config.

**Comment by snicoll** [comment]
we're talking about harmonizing here so we should do that regardless of the library implementation. I frankly prefer to go the `CacheManager` route and not expose things that are not strictly necessary. 

In the case of JCache, the environment may expose a `javax.cache.CacheManager` that we can wrap automatically. I don't think it would be wise to expose it as a `@Bean` by default (there is a `close` method on it so it would be disposed as well). My point is that currently if it is exposed as `@Bean` it _will_ be disposed and there is no way to disable that behaviour if the bean implements `Closeable`.

Long story short: it's not as easy as it sounds.

**Comment by wilkinsona** [comment]
Can't we annotate the bean method with `destroyMethod=""` to prevent `close` from being called?

**Comment by snicoll** [comment]
That's what I thought as well and learned it the hard way.  If the bean implements  `AutoCloseable` then we always call close (I am not saying it's a good thing, though).

**Comment by snicoll** [comment]
I think we should summarize this issue with whether or not we want to expose s the cache-specific infrastructure (on top of our standard `CacheManager`). If we want to do that, do we want to always do it? And do we want a common bean name for such a thing? 

Not sure I have answers to any of those questions. But what I do know for sure is that I want you to avoid to reconfigure thing just because Spring Boot does not expose what you need.

**Comment by snicoll** [comment]
I have also created an issue on the JSR-107 spec to clarify the lifecycle of the `javax.cache.CacheManager` that we would expose via this issue: https://github.com/jsr107/jsr107spec/issues/316

**Comment by artembilan** [comment]
@snicoll , thank you for the fix, but it doesn't address the main purpose of this ticket.
We need a separate `HazelcastConfiguration`. We aren't interested in Cache abstraction...
I see by your code, if I don't have `hazelcast-spring` dependency I won't get `HazelcastInstance` from this your caching stuff.

**Comment by snicoll** [comment]
The two first comment led me to think the contrary. Can you explain a bit more what you want? This is about auto-configuring the cache abstraction. If you don't care about that, maybe you shouldn't use it?
